python/guide/module/autograd.md

```diff
@@ -1,178 +1,286 @@
+
+With `autograd` it's simple, but there's one key difference compared to parameters: parameters are assumed to require gradients by default, non-parameters are not. We need to explicitly state that we require the gradient, and we do that by calling `.attach_grad()` on the `ndarray`. We can then access the gradient using `.grad` after the `backward` pass.
+
+As a simple example, let's take the case where $y=2x^2$ and use `autograd` to calculate gradient of $y$ with respect to $x$ at three different values of $x$. We could obviously work out the gradient by hand in this case as $dy/dx=4x$, but let's use this knowledge to check `autograd`. Given $x$ is an `ndarray` and not a `Parameter`, we need to call `x.attach_grad()`.
+
+```{.python .input  n=2}
+x = array([1, 2, 3])
+x.attach_grad()
 with autograd.record():
-    print(autograd.is_training())
-```
-
-In some cases, the same model behaves differently in the training and prediction
-modes (such as batch normalization). In other cases, some models may store more
-auxiliary variables to make computing gradients easier. We will cover these
-differences in detail in later chapters. For now, you need not worry about these
-details just yet.
-
-## Computing the Gradient of Python Control Flow
-
-One benefit
-of using automatic differentiation is that even if the computational graph of
-the function contains Python's control flow (such as conditional and loop
-control), we may still be able to find the gradient of a variable. Consider the
-following program:  It should be emphasized that the number of iterations of the
-loop (while loop) and the execution of the conditional judgment (if statement)
-depend on the value of the input `b`.
-
-```{.python .input  n=8}
-def f(a):
-    b = a * 2
-    while b.norm().asscalar() < 1000:
-        b = b * 2
-    if b.sum().asscalar() > 0:
-        c = b
+    y = 2 * x ** 2
+y.backward()
+print(x.grad)
+```
+
+## Advanced: Using Python control flow
+
+As mentioned before, one of the main advantages of `autograd` is the ability to automatically calculate gradients of dynamic graphs (i.e. graphs where the operators could be different on every forward pass). One example of this would be applying a tree structured recurrent network to parse a sentence using its parse tree. And we can use Python control flow operators to create a dynamic flow that depends on the data.
+
+We'll write a function as a toy example of a dynamic network. We'll add an `if` condition and a loop with a variable number of iterations, both of which will depend on the input data. Although these can now be used in static graphs (with conditional operators) it's still much more natural to use native control flow.
```

Comment:
I would add a link to this tutorial here: https://mxnet.incubator.apache.org/versions/master/tutorials/control_flow/ControlFlowTutorial.html

Reply:
thanks! updated.